virtualpdu/device_provider.py

```diff
@@ -0,0 +1,25 @@
+# Copyright 2016 Internap
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+class DeviceProvider(object):
+    def power_on(self, name):
+        raise NotImplementedError()
+
+    def power_off(self, name):
+        raise NotImplementedError()
+
```

Comment:
My plan was to handle this in the "core". The actual PDU would "receive" a REBOOT action, and that will end up doing power_off/power_on to the device driver. This is not unlike the current implementation in virtualdatacenter:

                if value == POWER_ON:
                    virsh.start(server_name)
                elif value == POWER_OFF:
                    virsh.stop(server_name)
                elif value == REBOOT:
                    virsh.stop(server_name)
                    virsh.start(server_name)

I do not think libvirt's "reboot" is the same because it might soft reboot the machine, which is not what we want here.